### PR TITLE
AP_HAL_Linux: Support "non-standard" baudrates on UART

### PR DESCRIPTION
**Pull request by nbertram**
The current glibc-based UART code using `termios.h` and `cfsetspeed()` isn't (usually) capable of setting baudrates outside the standard set. I discovered this on a Beaglebone Blue trying to use a ELRS (CRSF) RX which wants 416666 baud. The existing call to `cfsetspeed()` was failing silently in this case (I observed with `stty` that the port was remaining at 115200 throughout the RC protocol search).

Using the Linux-specific ioctl to set custom bauds using `CBAUDEX` fixed the issue. I also added some rudimentary error logging which wasn't present originally.

I note that the SITL `UART_utils.cpp` already uses this ioctl, though is using a different option to set a custom baud (`tc.c_cflag |= BOTHER`). I'm not sure if it's worth attempting a consolidation of Linux serial handling code across these contexts, open to suggestions.

### CODE REVIEW

File: libraries/AP_HAL_Linux/UARTDevice.cpp [hunk 1]
```diff
@@ -86,55 +87,89 @@ void UARTDevice::set_blocking(bool blocking)
...
 
-    if (_flow_control == flow_control_setting) {
+    if (ioctl(_fd, TCGETS2, &t) != 0) {
+        ::fprintf(stderr, "Failed to read serial options for %s - %s\n",
+                  _device_path, strerror(errno));
         return;
     }
 
-    tcgetattr(_fd, &t);
+    if (_flow_control == flow_control_setting) {
```
Comment (andyp1per) [inline libraries/AP_HAL_Linux/UARTDevice.cpp:144]: I think this check should be first

**Review by andyp1per — CHANGES_REQUESTED** [review]
Very nice! Just one change I think needed

File: libraries/AP_HAL_Linux/UARTDevice.cpp (hunk 1 above)
Comment (nbertram) [reply libraries/AP_HAL_Linux/UARTDevice.cpp]: Ahh yes, good catch. Updated. I also moved the declaration of the struct below that check, which wasn't the way it was before. Not sure if you prefer old C up-top variable declarations. If you do, I'll put it back where it was :smile:

**Review by andyp1per — APPROVED** [review]
Just needs to pass CI

**Comment by nbertram** [comment]
Sorted. Bloody musl.

**Comment by tridge** [comment]
assigned to @Williangalvani
this does make sense, just nice to get Willian to test it too

**Review by Williangalvani — APPROVED** [review]
works great! this actually also fixed the sbus issue I was facing.

**Pull request merged**